File: Python/Algorithms/SortingAlgorithms/InsertionSort.py
class InsertionSort:
    """ insertionSort Algorithm Implementation in Python 3

    arr : Unorded list
    output : Return list in ascending order.
    time complexity : O(n2)

    Example :
    >>> sort = InsertionSort
    >>> sort([4,2,6,5,9,8])
    [2, 4, 5, 6, 8, 9]"""

    def __init__(self):
        print("Insertion Sort Algorithm is Initialized")

    def __call__(self, arr):
        for i in range(1, len(arr)):
            key_item = arr[i]

            j = i - 1

            while j >= 0 and arr[j] > key_item:
                arr[j + 1] = arr[j]
                j -= 1

            arr[j + 1] = key_item

        return arr

File: C++/Algorithms/SortingAlgorithms/insertionSort.cpp
#include <bits/stdc++.h>
using namespace std;

/*
    Function to sort an array using Insertion Sort;
    Time Complexity : O(n*n);
    Space Complexity : O(1);
*/
void insertionSort(int arr[], int n) {

	for (int i = 1; i < n; i++) {
		int k = arr[i];
		int j = i - 1;
		while (j >= 0 and k < arr[j]) {
			arr[j + 1] = arr[j];
			j--;
		}
		arr[j + 1] = k;
	}

}

/*
    Utility Function to print elements of the array;
    Time Complexity : O(n);
    Space Complexity : O(1);
*/
void printArray(int arr[], int n) {

	for (int i = 0; i < n; i++) {
		cout << arr[i] << " ";
	}
	cout << endl;

}

int main() {

	int arr[] = {4, 7, 9, 1, 2, 3, 5};
	int n = sizeof(arr) / sizeof(arr[0]);

	insertionSort(arr, n);
	printArray(arr, n);

	return 0;

}